pkg/plugins/manifest.go

```diff
@@ -82,49 +84,130 @@ func readPluginManifest(body []byte) (*pluginManifest, error) {
+				return err
+			}
+
+			log.Debug("Handling symlink", "source", path, "target", finalPath)
+
+			info, err := os.Stat(finalPath)
+			if err != nil {
+				return err
+			}
+			if info.IsDir() {
+				return nil
+			}
+
+			if _, err := filepath.Rel(dpath, finalPath); err != nil {
+				return fmt.Errorf("symbolic link %q targets a file outside of the plugin directory: %q", path, finalPath)
+			}
+
+			if finalPath == manifestPath {
+				return nil
+			}
+
+			fi = info
+		}
+
+		if path == manifestPath {
+			return nil
+		}
+
+		// Ignore images
+		for _, sfx := range []string{"png", "gif", "svg"} {
+			if strings.HasSuffix(fi.Name(), fmt.Sprintf(".%s", sfx)) {
+				return nil
+			}
+		}
+
+		h := sha256.New()
+		f, err := os.Open(path)
+		if err != nil {
+			return err
+		}
+		defer f.Close()
+		if _, err := io.Copy(h, f); err != nil {
+			return err
+		}
+
+		relPath, err := filepath.Rel(dpath, path)
+		if err != nil {
+			return err
+		}
+		chksums[relPath] = fmt.Sprintf("%x", h.Sum(nil))
+
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	return chksums, nil
+}
+
 // getPluginSignatureState returns the signature state for a plugin.
-func getPluginSignatureState(log log.Logger, plugin *PluginBase) PluginSignature {
+func getPluginSignatureState(log log.Logger, plugin *PluginBase) (PluginSignature, error) {
 	log.Debug("Getting signature state of plugin", "plugin", plugin.Id)
 	manifestPath := path.Join(plugin.PluginDir, "MANIFEST.txt")
 
 	byteValue, err := ioutil.ReadFile(manifestPath)
 	if err != nil || len(byteValue) < 10 {
-		return PluginSignatureUnsigned
+		return PluginSignatureUnsigned, nil
 	}
 
 	manifest, err := readPluginManifest(byteValue)
 	if err != nil {
-		return PluginSignatureInvalid
+		return PluginSignatureInvalid, nil
 	}
 
 	// Make sure the versions all match
 	if manifest.Plugin != plugin.Id || manifest.Version != plugin.Info.Version {
-		return PluginSignatureModified
+		return PluginSignatureModified, nil
+	}
+
+	chksums, err := getChksums(plugin.PluginDir, manifestPath, log)
+	if err != nil {
+		return PluginSignatureInvalid, err
+	}
+	if len(chksums) != len(manifest.Files) {
+		log.Warn("Plugin's files differ from manifest", "plugin", plugin.Id)
+		return PluginSignatureModified, nil
 	}
 
 	// Verify the manifest contents
 	log.Debug("Verifying contents of plugin manifest", "plugin", plugin.Id)
 	for p, hash := range manifest.Files {
-		// Open the file
 		fp := path.Join(plugin.PluginDir, p)
-		f, err := os.Open(fp)
-		if err != nil {
-			return PluginSignatureModified
-		}
-		defer f.Close()
 
-		h := sha256.New()
-		if _, err := io.Copy(h, f); err != nil {
-			log.Warn("Couldn't read plugin file", "plugin", plugin.Id, "filename", fp)
-			return PluginSignatureModified
+		gotHash, ok := chksums[p]
+		if !ok {
+			log.Warn("Plugin file from manifest is missing", "plugin", plugin.Id, "filename", fp)
 		}
-		sum := hex.EncodeToString(h.Sum(nil))
-		if sum != hash {
+		if gotHash != hash {
```

Comment:
Your call as I'm not so sure on this. But since the zero value of both `gotHash` and `hash` is an empty string and we have continued from above, this could maybe too weak of a check if somehow one ends up with empty string equals empty string. But possibly in the context this is just be overly defensive programming.

Alternatively one could check that it isn't empty (assuming it never should be?), or maybe make it nullable (a pointer).  If empty string check would work that is probably simpler.

Reply:
@kylebrandt I guess an empty 256-bit checksum would be a bug? It should just be defensive programming to guard against this, but if we want to, we would also have to add these guards in grafana/build-pipeline, in the code to generate the manifest.